trans/src/Pact/Trans/Dec.hs

```diff
@@ -129,55 +135,74 @@ reduce_ n@N.Num { N.coefficient = c, N.exponent = e }
-  case evalArith (Op.exp (fromDecimal x :: Decimal)
+dec_exp :: TransGasParams -> D.Decimal -> TransResult (D.Decimal, Word64)
+dec_exp gp x =
+  case runArith (Op.exp (fromDecimal x :: Decimal)
                   :: Arith Prec N.RoundHalfEven Decimal)
-       extendedDefaultContext of
-    Left err -> error $ "exp error: " ++ show err
-    Right n -> toDecimal n
-
-dec_ln :: D.Decimal -> TransResult D.Decimal
-dec_ln x =
-  case evalArith (Op.ln (fromDecimal x :: Decimal)
+       (gasContext gp) of
+    (Left err, ctx) 
+      | exceptionSignal err == GasExceeded -> TransGasExceeded (ctxGas ctx)
+      | otherwise -> error $ "exp error: " ++ show err
+    (Right n, ctx) ->  (,ctxGas ctx) <$> toDecimal n
+
+dec_ln :: TransGasParams -> D.Decimal -> TransResult (D.Decimal, Word64)
+dec_ln gp x =
+  case runArith (Op.ln (fromDecimal x :: Decimal)
                   :: Arith Prec N.RoundHalfEven Decimal)
-       extendedDefaultContext of
-    Left err -> error $ "ln error: " ++ show err
-    Right n -> toDecimal n
-
-dec_log :: D.Decimal -> D.Decimal -> TransResult D.Decimal
-dec_log D.Decimal{ D.decimalPlaces = _, D.decimalMantissa = 0 } _ =
-  TransNumber 0.0
-dec_log b x =
-  case dec_ln x of
-    TransNumber x' ->
-      case dec_ln b of
-        TransNumber b' -> TransNumber (x' / b')
+       (gasContext gp) of
+    (Left err, ctx) 
+      | exceptionSignal err == GasExceeded -> TransGasExceeded (ctxGas ctx)
+      | otherwise -> error $ "ln error: " ++ show err
+    (Right n, ctx) -> (,ctxGas ctx) <$> toDecimal n
+
+dec_log :: TransGasParams -> D.Decimal -> D.Decimal -> TransResult (D.Decimal, Word64)
+dec_log gp D.Decimal{ D.decimalPlaces = _, D.decimalMantissa = 0 } _ =
+  TransNumber (0.0, _tGas gp)
+dec_log gp b x =
+  case dec_ln gp x of
+    TransNumber (x', g1) ->
+      case dec_ln (gp{_tGas=g1}) b of
+        TransNumber (b',g2) -> TransNumber (x' / b', g2)
         b' -> b'
     x' -> x'
 
-dec_pow :: D.Decimal -> D.Decimal -> TransResult D.Decimal
-dec_pow D.Decimal{ D.decimalPlaces = _, D.decimalMantissa = 0 }
+dec_pow :: TransGasParams -> D.Decimal -> D.Decimal -> TransResult (D.Decimal, Word64)
+dec_pow gp D.Decimal{ D.decimalPlaces = _, D.decimalMantissa = 0 }
         D.Decimal{ D.decimalPlaces = _, D.decimalMantissa = 0 } =
-  TransNumber 1.0
-dec_pow x y =
-  case evalArith (Op.power (fromDecimal x :: Decimal)
+  TransNumber (1.0, _tGas gp)
+dec_pow gp x y =
+  case runArith (Op.power (fromDecimal x :: Decimal)
                    (fromDecimal y :: Decimal)
                   :: Arith Prec N.RoundHalfEven Decimal)
-       extendedDefaultContext of
-    Left err -> error $ "pow error: " ++ show err
-    Right n -> toDecimal n
-
-dec_sqrt :: D.Decimal -> TransResult D.Decimal
-dec_sqrt x =
-  case evalArith (Op.squareRoot (fromDecimal x :: Decimal)
+       (gasContext gp) of
+    (Left err, ctx) 
+      | exceptionSignal err == GasExceeded -> TransGasExceeded (ctxGas ctx)
+      | otherwise -> error $ "ln error: " ++ show err
```

Comment:
this is `pow error`

Reply:
honestly we need to remove all uses of `error`

Reply:
why are we erroring here? That's a pretty hard kill. I'd prefer being in a monad here before using `error`.

Reply:
great question, we're erroring here because this was from john's pr. We shouldn't error anymore.